src/test/java/dk/kb/labsapi/ImageExportTest.java

```diff
@@ -42,6 +59,91 @@ public class ImageExportTest {
+
+            assertEquals(0, ImageExport.setUsableEndYear(2050),
+                    "Calling a mocked static with unsupported input should return null (= 0 as int is an Atomic)");
+        }
+    }
+
+    @Test
+    void testInstanceMocking() throws MalformedURLException {
+        // All mocking takes place inside the try-block
+        // We use Mockedconstruction as we want the standard constructor for ImageExport to be evaluated
+        try (MockedConstruction mocked = mockConstruction(ImageExport.class)) {
+            ImageExport mockedExport = ImageExport.getInstance();
+            when(mockedExport.convertPageUUID("foo")).thenReturn("bar", "bar2");
+            when(mockedExport.convertPageUUID("hello")).thenReturn("world");
+
+            // Perform unit tests using the mocked instance
+            assertEquals("bar", mockedExport.convertPageUUID("foo"),
+                    "Calling a mocked method with the first correct input should return the first expected output");
+            assertEquals("bar2", mockedExport.convertPageUUID("foo"),
+                    "Calling a mocked method a second time with the first correct input should return the second expected output");
+            assertEquals("bar2", mockedExport.convertPageUUID("foo"),
+                    "Calling a mocked method a third time with the first correct input should return the last output in the response chain");
+
+            assertEquals("world", mockedExport.convertPageUUID("hello"),
+                    "Calling a mocked method with the other correct input should return the expected output");
+
+            assertNull(mockedExport.convertPageUUID("zoo"),
+                    "Calling a mocked method with unexpected input should return null");
+        }
+    }
+
+
+    @Test
+    public void testMockedDownload() throws IOException {
+        URL testUrl = new URL("http://callisto.statsbiblioteket.dk/iipsrv/iipsrv.fcgi?FIF=/avis-show/symlinks/0/0/0/0/00005aff-ea53-46dd-bc90-0b0dd8917dbc.jp2&CVT=jpeg");
+        String testResponseString = "Image1";
+        byte[] testResponse = testResponseString.getBytes(StandardCharsets.UTF_8);
+
+        // Create export and spy on it
+        ImageExport export = ImageExport.getInstance();
+        ImageExport exportSpy = spy(export);
+
+        try {
+            // Define which method to stub
+            when(exportSpy.downloadSingleIllustration(testUrl)).thenReturn(testResponse);
+
+            ByteArrayOutputStream output = new ByteArrayOutputStream();
+
+            // Call method that we want to test, which uses the stubbed method above.
+            exportSpy.exportFullpages("pageUUID:doms_aviser_page:uuid:00005aff-ea53-46dd-bc90-0b0dd8917dbc", 1666, 1800, 1, output, "fullPage");
+
+            // Do tests on result
+            String content = output.toString(StandardCharsets.UTF_8);
+            assertEquals("Image1", content);
+            assertNotNull(output);
+        } catch (RuntimeException e){}
```

Comment:
What is the purpose of the try-catch here? If the `spy` needs to be closed somehow it should be in the catch.

Reply:
Sorry, but there seems to be an error in the test. The log message
```
16:08:14.915 [main] ERROR dk.kb.labsapi.ImageExport - Failed to download illustration from http://callisto.statsbiblioteket.dk/iipsrv/iipsrv.fcgi?FIF=/avis-show/symlinks/0/0/0/0/00005aff-ea53-46dd-bc90-0b0dd8917dbc.jp2&CVT=jpeg while reading bytes
```
indicates that an attempt is made.

I don't know why your method does not work, but after reading https://www.baeldung.com/mockito-spy#stubbing-a-spy I tried removing the try-catch and changing
```
    when(exportSpy.downloadSingleIllustration(testUrl)).thenReturn(testResponse);
```
to
```
    doReturn(testResponse).when(exportSpy).downloadSingleIllustration(testUrl);
```
This made the test fail in a new and seemingly better way :-)

Reply:
Yes! This is indeed a way better failure! It shows that the mock works and that I am asserting on a very low level here.

Reply:
So close now! Could you also mock the Solr call used by indirectly in the test method?